docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,80 @@
+status: accepted
+owner: s-urbaniak,bwplotka
+---
+
+### Related Tickets
+
+https://github.com/thanos-io/thanos/issues/1375
+
+## Summary
+
+Thanos allows a global query view for the Prometheus series. This means discovering, connecting various (often remote) "leafs" components and aggregating series data from them.
+
+Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it Thanos Ruler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy Rules API and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
```

Comment:
I wonder if it's really worth adding yet another flag. Because we know the StoreAPI types already perhaps it would be cleaner to use that information? :thinking: What does everyone else think about this idea?

Reply:
I think I vote for another flag. Yes, it's +2 flags (probably we need `-strict` as well) but really it's just a flag - it has it's --help and it's readable. I think we need separation here. What if you want to use some server as Store API WITHOUT using it as RuleAPI? What if the opposite? I think different API needs different flag. (: 


Happy to hear other opinions.

Reply:
In what use-cases you could want only to retrieve one or the other? :P Can't think of any off the top of my head. Another option would be to rename `--store` & friends to something like `--node` and then use gRPC reflection to know what services are exposed to us by the specified servers (https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection).

Nonetheless, we can go with the very explicit option that you are proposing here. I'm just afraid of the configuration complexity & verbosity. I think on our Slack we've talked a bit how we could implement another API for something else but I couldn't find the discussion now. So, we would get another pair of parameters in such a case? If not now then in the near future I think these should definitely be joined together.

Reply:
> -store & friends to something like --node and then use gRPC reflection to know what services are exposed to us by the specified servers (https://github.com/grpc/grpc-go/blob/master/Documentation/server-reflection-tutorial.md#enable-server-reflection).

I was thinking initially that it's a good idea, but IMO it's not. It's better, especially for initial iteration to allow user for explicit configuration - always simpler. WDYT? What's the harm here?

Reply:
yes the other one would be `TargetsAPI`

Reply:
It is fair to talk about this now.. but I think explict `--target` would be needed as well